src/main/java/org/prebid/server/bidder/adocean/AdoceanBidder.java

```diff
@@ -170,38 +207,55 @@ private String buildUrl(String impid, ExtImpAdocean extImpAdocean, String consen
             uriBuilder.addParameter("hcuserid", user.getBuyeruid());
         }
 
+        final List<String> sizeValues = test != null && test == 1
+                ? setSlaveSizesParam(slaveSizes, true)
+                : setSlaveSizesParam(slaveSizes, false);
+
+        if (CollectionUtils.isNotEmpty(sizeValues)) {
+            uriBuilder.addParameter("aosspsizes", String.join("-", sizeValues));
+        }
+
         return uriBuilder.toString();
     }
 
+    private List<String> setSlaveSizesParam(Map<String, String> slaveSizes, boolean orderByKey) {
+        final Set<String> slaveIDs = orderByKey ? new TreeSet<>(slaveSizes.keySet()) : slaveSizes.keySet();
+
+        return slaveIDs.stream()
+                .filter(slaveId -> StringUtils.isNotBlank(slaveSizes.get(slaveId)))
+                .map(rawSlaveID -> String.format("%s~%s", rawSlaveID.replaceFirst("adocean", ""),
+                        slaveSizes.get(rawSlaveID)))
+                .collect(Collectors.toList());
+    }
+
     private static MultiMap getHeaders(BidRequest request) {
         final MultiMap headers = HttpUtil.headers();
         if (request.getDevice() != null) {
-            HttpUtil.addHeaderIfValueIsNotEmpty(headers, HttpUtil.USER_AGENT_HEADER.toString(),
-                    request.getDevice().getUa());
-
-            if (StringUtils.isNotBlank(request.getDevice().getIp())) {
-                HttpUtil.addHeaderIfValueIsNotEmpty(headers, HttpUtil.X_FORWARDED_FOR_HEADER,
-                        request.getDevice().getIp());
-            } else if (StringUtils.isNotBlank(request.getDevice().getIpv6())) {
-                HttpUtil.addHeaderIfValueIsNotEmpty(headers, HttpUtil.X_FORWARDED_FOR_HEADER,
-                        request.getDevice().getIpv6());
-            }
+            addHeader(headers, "User-Agent", request.getDevice().getUa());
```

Comment:
Check for nulls in values for headers.

Reply:
method addHeader contains StringUtils.isNotBlank(value)

Reply:
fair